### ISSUE TITLE
is there support for custom Config rule with lambda running in central account

### ISSUE THREAD

**Issue by tannaga** [cloud-custodian/cloud-custodian#3607]
When creating Custom Config Rule, does Cloud Custodian provide an ability to specify the arn of the lambda from a different account (central account)? if yes, an example would be great..

**Comment by kapilt** [comment]
No, there isn't support for that.. But given custodian is creating the lambda, config rule, ie. its managing everything, What is the benefit of the centralized lambda?  the central  ends up being a large blast radius and also has scaling limitations (due to lambda permission size).

**Comment by tannaga** [comment]
@kapilt i understand the limitation on lambda invoke permission at present, but if they start supporting principal org id, it will be solved
I agree on the downside of central lambda but there are some of advantages too
- Easier to set up  logging/monitoring/alerts
- Lesser footprints in target accounts. Makes it easier to ensure immutability of policy definition. So in this case it would be the config rule alone to monitor.
- In case of a buggy code, easier to reprocess the request with the help of DLQ in central account design.

Since AWS Config Rule does provide ability to add cross account lambda, I feel providing it as an option will surely help.  It would then be up to teams to weigh the pros and cons based on their requirements.

**Comment by kapilt** [comment]
org id would solve one aspect on the limitation, ie make it possible, but i don't think it addresses the larger concern, that this will cause more problems for users rather then less. ie. larger blast radius, harder to debug, etc. I'm willing to be convinced, but i'm not here anything compelling, i think people think that one is going to be easier, but that is often a false dichotomy imo due to not having experience troubleshooting problems with the one. Many small are often much easier to reason about when an individual one has problems. Custodian tries to automate the management of the many so that the additional operation overhead of them is negligible. reviewing your proposed advantages one by one.

monitoring/logging/alerts - i actually regard the logging and metrics to be harder in centralized setups, outside of a trivial alert, because actually debugging anything is much more complicated due to the multi account event stream and execution logs and metrics. 

footprint, the reality is that once you have to do x in account, doing x + 1 is a nominal cost imo. the process to provision and ensure the footprints 'immutability' or presence is the same, you run the custodian cli.

dlq - i haven't seen or heard of a single person doing event re-processing with custodian, we do support dlq configuration, but thats for manual introspection. Also you can do centralized dlqs today, for lambda policies across accounts.

so, let's say we wanted to go ahead and support this.. what do you think it should look like to spell in policy configuration. ie example config. also is this something your willing to work on/contribute?

**Comment by tannaga** [comment]
I think/feel that blast radius will be actually less/controlled. Let's say  a bug is introduced with a new release, we will be able to quickly identify and revert to the older versions at a single place vs updating it in ~1500 accounts *16 regions.  
I should have been clear with the DLQ earlier. What i intend to mean is that if the events are stored in a queue, it will help us replay the event and process the message if and when needed.

This would be similar to having a policy with event bus configured on Cloudwatch events in target accounts and then processing the events in the central account.

Policy configuration may look like this.. one lambda for each rule.
```
policies:
  - name: my-custom-rule
    description: |
      creating a custom rule
    resource: s3
    mode:
      type: config-rule
      lambda_arn: arn:aws:lambda:us-west-2:0123456789:function:custom-rule-processor
```
I can surely try to contribute here. 
WDYT?

**Comment by kapilt** [comment]
I think your confusing operational/sre terms. Blast radius is about scope/size of a change, what your describing is time to recovery which is an orthogonal concept. If you have 1500 accounts and 16 regions, doing an atomic change to all of them is a large blast radius. Operational excellence would have you break that change into smaller sizes, ie. canary them in a dev/qa account before rolling out the change more widely, or doing a rolling upgrade across sets of accounts. Again custodian focuses on providing operational tools that enable handling the many well. using c7n-org (parallel policy execution/deployment across accounts and regions) and your recovery scenario, would take under 5 minutes.

Wrt to your example policy, in the absence of org id support in lambda permissions, you'd also need to specify all of the account ids of the target accounts in the policy or the permission would need to remove the account id qualifier.

**Comment by kapilt** [comment]
Also wrt to DLQ and config rules, there seems to be a lack of understanding of how config evaluations work. They are not replayable the same way a cloudwatch event would be. Config invocations include a time limited token that is used to mark a resource as (non) compliant. Simple replay of a config invocation event would generate an error on invalid token. The proper way to reinvoke a config evaluation is to use the service api (in each account/region) to generate a re-evaluation.

**Comment by tannaga** [comment]
Thanks for throwing in ideas. I will explore cfn-org as well.. 
My point of view is that.. since many of our policies are event based canary deployment does not fit well in this context because you have to deploy in few accounts and wait for the trigger to happen to take further decision. Sure we can always mock the event, but that would be similar to what we use in our test env.
For the DLQ, we get handle of the account, region and the config rule which will help us [re-evaluate](https://docs.aws.amazon.com/cli/latest/reference/configservice/start-config-rules-evaluation.html)

I think i got my original question answered... I will give a shot in my local repo and see how it goes.
Thanks for your time.
